R/utils.R

```diff
@@ -0,0 +1,325 @@
+  init <- parsed_string[1, ]
+
+  # Trim leading and trailing commas
+  init$value <- trimws(init$value, "both")
+  init$value <- trimws(init$value, "both", whitespace = "\"")
+
+  # If only one string return
+  if (nrow(parsed_string) == 1) {
+    return(init)
+  }
+  for (j in seq(2, nrow(parsed_string))) {
+
+    # Search if need to replace and concat
+    row <- parsed_string[j, ]
+
+    value <- replace_string_and_concat(
+      row$value,
+      init$name_string,
+      init$value
+    )
+
+    row$value <- trimws(value, "both", "\"")
+
+    init <- rbind(init, row)
+  }
+
+  return(init)
+}
+
+#' Parse a single entry
+#'
+#' Returns an entry from a .bib file parsed into an R object
+#'
+#' @noRd
+parse_single_entry <- function(init, end, lines, map_string_end) {
+  entry_lines <- lines[seq(init, end)]
+
+  # Skipping empty lines
+  entry_lines <- entry_lines[entry_lines != ""]
+
+  # Guess lines outside of the entry
+
+  guess_eoentry <- max(grep("\\}$", entry_lines))
+
+  entry_lines <- entry_lines[seq(1, guess_eoentry)]
+
+  # Collapse to single line
+  entry_collapsed <- paste0(entry_lines, collapse = "\n")
+
+  # Check balanced braces
+  map_braces <- check_balanced_braces(entry_collapsed, init)
+
+  # Extract string from beginning to first balanced brace
+
+  lastbrace <- min(map_braces[map_braces$cum_flag == 0, ]$pos)
+
+  # This would remove the last brace, it is what we want
+  cleaned_entry <- trimws(substr(entry_collapsed, 1, lastbrace - 1))
+
+  # Identify fields
+  # Safest pattern I found:
+  # ,<space or newline>singleword<space or newline>=
+  # E.g:
+  # ...}, author={...}
+  # ...",\ntitle ="
+  # etc.
+  # As field entry could have -, _ or : need to replace when identifying the
+  # fields. This improve the regex matching pattern
+
+  cleaned_entry_sub <-
+    gsub("(?<=[[:alnum:]])_|(?<=[[:alnum:]])-|(?<=[[:alnum:]]):", "x",
+         cleaned_entry,
+         perl = TRUE
+    )
+
+  # Protect commas on brackets to avoid error on splitting
+  protected <- gsub(",(?![^\\}]*(\\{|$))", "|",
+                    cleaned_entry_sub,
+                    perl = TRUE
+  )
+
+  posfields <- unlist(gregexpr(",\\s*\\w+\\s*=", protected))
+  # Add init and last pos
+  posfields <- sort(unique(c(1, posfields, nchar(cleaned_entry))))
+
+  formatted_bib <- lapply(seq_len(length(posfields) - 1), function(x) {
+    str <- substr(cleaned_entry, posfields[x], posfields[x + 1])
+
+    # Cleanup leading and trailing comma and blanks
+    str <- trimws(gsub("^,|,$", "", str))
+  })
+
+  formatted_bib <- unlist(formatted_bib)
+
+  # Each field is now on a single line
+
+  # # Extract type of entry and key
+  # String to first comma
+  entry_key <- formatted_bib[1]
+
+  entry_key <- gsub("@", "", unlist(strsplit(entry_key, "\\{")))
+  entry_name <- trimws(entry_key[1])
+  key_name <- trimws(entry_key[2])
+
+  # Now treat fields
+
+  fields <- formatted_bib[-1]
+
+  treat_fields <- lapply(seq_len(length(fields)), function(x) {
```

Comment:
Since we already switch `fields` over to a vector, let's take advantage of `seq_along()` again. 

```suggestion
  treat_fields <- lapply(seq_along(fields), function(x) {
```